File: backend/src/main/java/com/kredent/backend/service/CertificateService.java
package com.kredent.backend.service;

import com.kredent.backend.dto.CertificateMetadataRequest;
import com.kredent.backend.dto.CertificateResponse;
import com.kredent.backend.dto.IssueBlockchainRequest;
import com.kredent.backend.dto.PageResponse;
import com.kredent.backend.dto.RevokeBlockchainRequest;
import com.kredent.backend.dto.UpdateCertificateStatusRequest;
import com.kredent.backend.entity.ActorType;
import com.kredent.backend.entity.Admin;
import com.kredent.backend.entity.Certificate;
import com.kredent.backend.entity.CertificateStatus;
import com.kredent.backend.entity.Student;
import com.kredent.backend.repository.CertificateRepository;
import com.kredent.backend.repository.StudentRepository;
import com.kredent.backend.util.FileValidationUtil;
import com.kredent.backend.util.HashUtil;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.server.ResponseStatusException;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Class-level @Transactional matters here specifically because Certificate's
 * student/issuedByAdmin associations are lazy (see entity comments): mapping
 * to CertificateResponse touches non-id fields on those proxies (fullName,
 * usn), which requires an open Hibernate session for the whole method, not
 * just for the repository call. This is a service-layer/JPA concern, not a
 * change to authentication.
 */
@Service
@Transactional
public class CertificateService {

    private final CertificateRepository certificateRepository;
    private final StudentRepository studentRepository;
    private final CurrentUserService currentUserService;
    private final AuditLogService auditLogService;
    private final SupabaseStorageService storageService;
    private final BlockchainVerificationService blockchainVerificationService;
    private final PdfStampingService pdfStampingService;
    private final long maxFileSizeBytes;
    private final String frontendBaseUrl;

    public CertificateService(
            CertificateRepository certificateRepository,
            StudentRepository studentRepository,
            CurrentUserService currentUserService,
            AuditLogService auditLogService,
            SupabaseStorageService storageService,
            BlockchainVerificationService blockchainVerificationService,
            PdfStampingService pdfStampingService,
            @Value("${app.certificate.max-file-size-mb:5}") long maxFileSizeMb,
            @Value("${app.frontend-base-url}") String frontendBaseUrl) {
        this.certificateRepository = certificateRepository;
        this.studentRepository = studentRepository;
        this.currentUserService = currentUserService;
        this.auditLogService = auditLogService;
        this.storageService = storageService;
        this.blockchainVerificationService = blockchainVerificationService;
        this.pdfStampingService = pdfStampingService;
        this.maxFileSizeBytes = maxFileSizeMb * 1024 * 1024;
        this.frontendBaseUrl = frontendBaseUrl;
    }

    public CertificateResponse createMetadata(CertificateMetadataRequest request) {
        Student student = studentRepository.findById(request.getStudentId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Student not found"));
        Admin admin = currentUserService.getCurrentAdmin();

        Certificate certificate = new Certificate();
        certificate.setCertificateNumber(generateCertificateNumber(request.getYearOfCompletion()));
        certificate.setStudent(student);
        certificate.setIssuedByAdmin(admin);
        certificate.setDegreeName(request.getDegreeName());
        // Always the student's own department — never independently chosen — so a certificate
        // can never disagree with the student record it belongs to (see CertificateMetadataRequest).
        certificate.setDepartment(student.getDepartment());
        certificate.setYearOfCompletion(request.getYearOfCompletion());
        certificate.setStatus(CertificateStatus.PENDING_MINT);
        certificateRepository.save(certificate);

        auditLogService.record(
                ActorType.ADMIN,
                admin.getId(),
                "CERTIFICATE_METADATA_ISSUED",
                "CERTIFICATE",
                certificate.getId().toString(),
                Map.of("certificateNumber", certificate.getCertificateNumber(), "studentId", student.getId())
        );

        return CertificateResponse.from(certificate);
    }

    public CertificateResponse getById(UUID id) {
        Certificate certificate = findOrThrow(id);
        assertCanView(certificate);
        return CertificateResponse.from(certificate);
    }

    public PageResponse<CertificateResponse> getOwnCertificates(Pageable pageable) {
        Student student = currentUserService.getCurrentStudent();
        Page<Certificate> page = certificateRepository.findByStudentId(student.getId(), pageable);
        return PageResponse.from(page, CertificateResponse::from);
    }

    /** Admin registry view — all certificates, optionally filtered by a search term (student name/USN/certificate number). */
    public PageResponse<CertificateResponse> listAll(String search, Pageable pageable) {
        Page<Certificate> page = (search == null || search.isBlank())
                ? certificateRepository.findAll(pageable)
                : certificateRepository.search(search.trim(), pageable);
        return PageResponse.from(page, CertificateResponse::from);
    }

    /**
     * Department-oriented Certificate Registry view — department, graduation year, and status
     * are each optional (null means "any"), combined with the same free-text search as listAll()
     * above. All filtering happens in SQL (CertificateRepository.searchFiltered), so this never
     * pulls more than one page of rows into memory regardless of how many certificates exist.
     */
    public PageResponse<CertificateResponse> listFiltered(
            String department, Integer year, CertificateStatus status, String search, Pageable pageable) {
        String query = search == null ? "" : search.trim();
        Page<Certificate> page = certificateRepository.searchFiltered(department, year, status, query, pageable);
        return PageResponse.from(page, CertificateResponse::from);
    }

    public PageResponse<CertificateResponse> getCertificatesForStudent(Long studentId, Pageable pageable) {
        if (!studentRepository.existsById(studentId)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Student not found");
        }
        Page<Certificate> page = certificateRepository.findByStudentId(studentId, pageable);
        return PageResponse.from(page, CertificateResponse::from);
    }

    /**
     * Attaches a PDF to an already-created certificate metadata row (see
     * createMetadata above). Kept as a separate step deliberately — it
     * doesn't touch the existing metadata-creation flow at all.
     */
    public CertificateResponse uploadFile(UUID id, MultipartFile file) {
        Certificate certificate = findOrThrow(id);
        validateFile(file);

        byte[] content;
        try {
            content = file.getBytes();
        } catch (IOException e) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Could not read the uploaded file");
        }

        if (!FileValidationUtil.looksLikePdf(content)) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
                    "The file does not look like a valid PDF (failed signature check)");
        }

        // Phase 11: stamp a QR code (linking to the public verify page for this certificate)
        // onto the PDF BEFORE anything below touches it. Everything that follows — hashing,
        // dedupe check, storage — operates on the STAMPED bytes, not the raw upload. That
        // ordering is what makes the QR forgery-resistant: it becomes part of the exact file
        // whose hash gets stored and later minted on-chain, so a copy of the QR pasted into a
        // different PDF produces bytes that no longer hash-match (caught by
        // PublicVerificationService.verifyPdf).
        String verificationUrl = frontendBaseUrl + "/verify/" + certificate.getCertificateNumber();
        byte[] stampedContent;
        try {
            stampedContent = pdfStampingService.stampQrCode(content, verificationUrl);
        } catch (RuntimeException e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
                    "Could not generate the verification QR code for this certificate", e);
        }

        // The hash is computed from the exact stamped bytes that will actually be stored and
        // served — this is the PDF's real fingerprint, independent of where it's stored.
        String fileHash = HashUtil.sha256Hex(stampedContent);

        certificateRepository.findByFileHash(fileHash).ifPresent(existing -> {
            if (!existing.getId().equals(id)) {
                throw new ResponseStatusException(HttpStatus.CONFLICT,
                        "This exact PDF has already been uploaded for another certificate (" + existing.getCertificateNumber() + ")");
            }
        });

        String sanitizedUsn = certificate.getStudent().getUsn().replaceAll("[^A-Za-z0-9]", "");
        String objectPath = "certificates/" + sanitizedUsn + "/" + UUID.randomUUID() + ".pdf";

        String fileUrl = storageService.upload(objectPath, stampedContent, FileValidationUtil.PDF_CONTENT_TYPE);

        certificate.setStoragePath(objectPath);
        certificate.setFileUrl(fileUrl);
        certificate.setOriginalFilename(file.getOriginalFilename());
        certificate.setFileSizeBytes((long) stampedContent.length);
        certificate.setMimeType(FileValidationUtil.PDF_CONTENT_TYPE);
        certificate.setFileHash(fileHash);
        certificate.setUploadedAt(LocalDateTime.now());
        certificateRepository.save(certificate);

        Admin admin = currentUserService.getCurrentAdmin();
        auditLogService.record(
                ActorType.ADMIN,
                admin.getId(),
                "CERTIFICATE_FILE_UPLOADED",
                "CERTIFICATE",
                id.toString(),
                Map.of(
                        "originalFilename", String.valueOf(file.getOriginalFilename()),
                        "fileSizeBytes", stampedContent.length,
                        "fileHash", fileHash
                )
        );

        return CertificateResponse.from(certificate);
    }

    public CertificateFile downloadFile(UUID id) {
        Certificate certificate = findOrThrow(id);
        assertCanView(certificate);

        if (certificate.getStoragePath() == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "No file has been uploaded for this certificate yet");
        }

        byte[] content = storageService.download(certificate.getStoragePath());
        String filename = certificate.getOriginalFilename() != null
                ? certificate.getOriginalFilename()
                : certificate.getCertificateNumber() + ".pdf";
        String mimeType = certificate.getMimeType() != null ? certificate.getMimeType() : FileValidationUtil.PDF_CONTENT_TYPE;

        return new CertificateFile(content, filename, mimeType);
    }

    public void deleteMetadata(UUID id) {
        Certificate certificate = findOrThrow(id);
        Admin admin = currentUserService.getCurrentAdmin();

        boolean hadFile = certificate.getStoragePath() != null;
        Map<String, Object> metadata = Map.of(
                "certificateNumber", certificate.getCertificateNumber(),
                "studentId", certificate.getStudent().getId(),
                "fileDeleted", hadFile
        );

        if (hadFile) {
            storageService.delete(certificate.getStoragePath());
        }
        certificateRepository.delete(certificate);

        auditLogService.record(
                ActorType.ADMIN,
                admin.getId(),
                "CERTIFICATE_METADATA_DELETED",
                "CERTIFICATE",
                id.toString(),
                metadata
        );
    }

    private void validateFile(MultipartFile file) {
        if (file == null || file.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "No file was uploaded, or the file is empty");
        }
        if (!FileValidationUtil.isDeclaredPdf(file.getContentType())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Only PDF files are accepted");
        }
        if (file.getSize() > maxFileSizeBytes) {
            throw new ResponseStatusException(HttpStatus.PAYLOAD_TOO_LARGE,
                    "File exceeds the maximum allowed size of " + (maxFileSizeBytes / (1024 * 1024)) + "MB");
        }
    }

    public CertificateResponse updateStatus(UUID id, UpdateCertificateStatusRequest request) {
        Certificate certificate = findOrThrow(id);

        if (request.getStatus() == CertificateStatus.REVOKED && (request.getReason() == null || request.getReason().isBlank())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "A reason is required to revoke a certificate");
        }
        // Once a certificate is minted on-chain, its status must stay in sync with the smart
        // contract — this plain endpoint can't do that, so it defers to revokeOnBlockchain()
        // below, which actually revokes on-chain first. Certificates that never got minted
        // (still PENDING_MINT / MINT_FAILED) keep working through this endpoint exactly as
        // they did in Phase 2 — nothing changes for that case.
        if (request.getStatus() == CertificateStatus.REVOKED && certificate.getStatus() == CertificateStatus.MINTED) {
            throw new ResponseStatusException(HttpStatus.CONFLICT,
                    "This certificate is minted on-chain — use the blockchain revoke action so the on-chain record stays in sync");
        }

        CertificateStatus previousStatus = certificate.getStatus();
        certificate.setStatus(request.getStatus());

        if (request.getStatus() == CertificateStatus.REVOKED) {
            certificate.setRevokedAt(LocalDateTime.now());
            certificate.setRevokedReason(request.getReason());
        } else {
            certificate.setRevokedAt(null);
            certificate.setRevokedReason(null);
        }
        certificateRepository.save(certificate);

        Admin admin = currentUserService.getCurrentAdmin();
        Map<String, Object> metadata = new HashMap<>();
        metadata.put("from", previousStatus);
        metadata.put("to", request.getStatus());
        if (request.getReason() != null) {
            metadata.put("reason", request.getReason());
        }
        auditLogService.record(
                ActorType.ADMIN,
                admin.getId(),
                "CERTIFICATE_STATUS_UPDATED",
                "CERTIFICATE",
                id.toString(),
                metadata
        );

        return CertificateResponse.from(certificate);
    }

    /**
     * Records a blockchain credential that the admin's MetaMask has already issued on-chain
     * (see frontend/src/services/blockchainService.js -> issueCredentialOnChain). This method
     * does not talk to MetaMask and does not submit any transaction — it independently verifies
     * the reported transaction really happened (BlockchainVerificationService), cross-checks it
     * against data we already had (student wallet, certificate hash), and only then saves it.
     */
    public CertificateResponse issueOnBlockchain(UUID id, IssueBlockchainRequest request) {
        Certificate certificate = findOrThrow(id);

        if (certificate.getFileHash() == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
                    "Upload the certificate PDF (and generate its hash) before issuing on the blockchain");
        }
        if (certificate.getStatus() == CertificateStatus.MINTED) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "This certificate has already been issued on the blockchain");
        }
        if (certificate.getStatus() == CertificateStatus.REVOKED) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "This certificate has been revoked and cannot be reissued");
        }

        String studentWallet = certificate.getStudent().getWalletAddress();
        if (studentWallet == null || studentWallet.isBlank()) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "This student does not have a system-managed wallet on file");
        }
        if (!studentWallet.equalsIgnoreCase(request.getStudentWalletAddress())) {
            throw new ResponseStatusException(HttpStatus.CONFLICT,
                    "The wallet address used on-chain does not match this student's wallet on file");
        }
        if (!certificate.getFileHash().equalsIgnoreCase(request.getCertificateHash())) {
            throw new ResponseStatusException(HttpStatus.CONFLICT,
                    "The certificate hash used on-chain does not match the stored SHA-256 hash");
        }

        blockchainVerificationService.verifyContractTransaction(request.getTransactionHash());

        certificate.setWalletAddress(studentWallet.toLowerCase());
        certificate.setTokenId(request.getTokenId());
        certificate.setContractAddress(request.getContractAddress());
        certificate.setTxHash(request.getTransactionHash());
        certificate.setStatus(CertificateStatus.MINTED);
        certificate.setMintedAt(LocalDateTime.now());
        certificateRepository.save(certificate);

        Admin admin = currentUserService.getCurrentAdmin();
        auditLogService.record(
                ActorType.ADMIN,
                admin.getId(),
                "CERTIFICATE_BLOCKCHAIN_ISSUED",
                "CERTIFICATE",
                id.toString(),
                Map.of(
                        "tokenId", String.valueOf(request.getTokenId()),
                        "transactionHash", request.getTransactionHash(),
                        "studentWalletAddress", studentWallet
                )
        );

        return CertificateResponse.from(certificate);
    }

    /**
     * Records an on-chain revocation the admin's MetaMask has already submitted (see
     * blockchainService.revokeCredentialOnChain). Only a currently MINTED certificate can be
     * revoked this way — a certificate that was never minted just uses the plain
     * updateStatus() path above, unchanged from Phase 2.
     */
    public CertificateResponse revokeOnBlockchain(UUID id, RevokeBlockchainRequest request) {
        Certificate certificate = findOrThrow(id);

        if (certificate.getStatus() != CertificateStatus.MINTED) {
            throw new ResponseStatusException(HttpStatus.CONFLICT,
                    "Only a certificate that is currently MINTED on-chain can be revoked on-chain");
        }
        if (certificate.getTokenId() == null || !certificate.getTokenId().equals(request.getTokenId())) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "Token ID does not match this certificate's on-chain record");
        }

        blockchainVerificationService.verifyContractTransaction(request.getTransactionHash());

        certificate.setStatus(CertificateStatus.REVOKED);
        certificate.setRevokedAt(LocalDateTime.now());
        certificate.setRevokedReason(request.getReason());
        certificateRepository.save(certificate);

        Admin admin = currentUserService.getCurrentAdmin();
        auditLogService.record(
                ActorType.ADMIN,
                admin.getId(),
                "CERTIFICATE_BLOCKCHAIN_REVOKED",
                "CERTIFICATE",
                id.toString(),
                Map.of(
                        "tokenId", String.valueOf(request.getTokenId()),
                        "transactionHash", request.getTransactionHash(),
                        "reason", request.getReason()
                )
        );

        return CertificateResponse.from(certificate);
    }

    private Certificate findOrThrow(UUID id) {
        return certificateRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Certificate not found"));
    }

    private void assertCanView(Certificate certificate) {
        if (currentUserService.isAdmin()) {
            return;
        }
        if (currentUserService.isStudent()) {
            Student student = currentUserService.getCurrentStudent();
            if (certificate.getStudent().getId().equals(student.getId())) {
                return;
            }
        }
        throw new ResponseStatusException(HttpStatus.FORBIDDEN, "You do not have access to this certificate");
    }

    private String generateCertificateNumber(int year) {
        String suffix = UUID.randomUUID().toString().replace("-", "").substring(0, 8).toUpperCase();
        return "SKC-" + year + "-" + suffix;
    }
}

File: frontend/src/services/certificateService.js
import { getAuthToken, parseErrorMessage } from '../utils/http'

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8081'

function authHeaders(extra = {}) {
  const token = getAuthToken()
  return {
    ...extra,
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
  }
}

// GET /api/admin/students?search=... — used to populate the student picker on Issue Degree.
export async function searchStudents(search = '') {
  const params = new URLSearchParams({ size: '50' })
  if (search) params.set('search', search)
  const res = await fetch(`${API_BASE}/api/admin/students?${params.toString()}`, {
    headers: authHeaders(),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json()
}

// POST /api/admin/certificates — creates the certificate metadata row (no file yet).
export async function createCertificate(payload) {
  const res = await fetch(`${API_BASE}/api/admin/certificates`, {
    method: 'POST',
    headers: authHeaders({ 'Content-Type': 'application/json' }),
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json()
}

// POST /api/admin/certificates/{id}/upload — attaches the PDF, which is where the
// backend computes the real SHA-256 hash and stores the file in Supabase Storage.
export async function uploadCertificateFile(certificateId, file) {
  const formData = new FormData()
  formData.append('file', file)
  const res = await fetch(`${API_BASE}/api/admin/certificates/${certificateId}/upload`, {
    method: 'POST',
    headers: authHeaders(), // no Content-Type — the browser sets the multipart boundary
    body: formData,
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json()
}

// GET /api/admin/certificates?search=&department=&year=&status=&page=&size= — the Certificate
// Registry. department/year/status are each optional; when any is set, filtering happens
// server-side (CertificateService.listFiltered on the backend) so this never needs to fetch more
// than one page of rows regardless of how many certificates exist.
export async function listCertificates({ search = '', department = '', year = '', status = '', page = 0, size = 20 } = {}) {
  const params = new URLSearchParams({ page: String(page), size: String(size) })
  if (search) params.set('search', search)
  if (department) params.set('department', department)
  if (year) params.set('year', String(year))
  if (status) params.set('status', status)
  const res = await fetch(`${API_BASE}/api/admin/certificates?${params.toString()}`, {
    headers: authHeaders(),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json()
}

// GET /api/student/certificates?page=&size= — only the logged-in student's own certificates.
// Ownership is enforced entirely server-side (the STUDENT-scoped JWT identifies the student —
// nothing here sends a student ID for the backend to trust).
export async function getMyCertificates({ page = 0, size = 100 } = {}) {
  const params = new URLSearchParams({ page: String(page), size: String(size) })
  const res = await fetch(`${API_BASE}/api/student/certificates?${params.toString()}`, {
    headers: authHeaders(),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json()
}

// GET /api/certificates/{id} — reachable by any authenticated user, but the backend only returns
// the certificate if it belongs to the requesting student (or the caller is an admin) — see
// CertificateService.assertCanView. A student can never fetch another student's certificate by
// changing this ID.
export async function getCertificateById(certificateId) {
  const res = await fetch(`${API_BASE}/api/certificates/${certificateId}`, {
    headers: authHeaders(),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json()
}

// PATCH /api/admin/certificates/{id}/status — used for revoking (reason required by the backend).
export async function updateCertificateStatus(certificateId, status, reason) {
  const res = await fetch(`${API_BASE}/api/admin/certificates/${certificateId}/status`, {
    method: 'PATCH',
    headers: authHeaders({ 'Content-Type': 'application/json' }),
    body: JSON.stringify({ status, reason }),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json()
}

// GET /api/certificates/{id}/download — streams the PDF; triggers a browser download.
export async function downloadCertificateFile(certificateId, suggestedFilename) {
  const res = await fetch(`${API_BASE}/api/certificates/${certificateId}/download`, {
    headers: authHeaders(),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  const blob = await res.blob()
  const url = window.URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = url
  link.download = suggestedFilename || 'certificate.pdf'
  document.body.appendChild(link)
  link.click()
  link.remove()
  window.URL.revokeObjectURL(url)
}

// POST /api/admin/certificates/{id}/blockchain/issue — called AFTER MetaMask has already
// signed and mined the on-chain issueCredential() transaction (see blockchainService.js). The
// backend independently re-verifies the transaction before saving any of this.
export async function issueCertificateBlockchain(certificateId, payload) {
  const res = await fetch(`${API_BASE}/api/admin/certificates/${certificateId}/blockchain/issue`, {
    method: 'POST',
    headers: authHeaders({ 'Content-Type': 'application/json' }),
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json()
}

// POST /api/admin/certificates/{id}/blockchain/revoke — same idea, for a revokeCredential() transaction.
export async function revokeCertificateBlockchain(certificateId, payload) {
  const res = await fetch(`${API_BASE}/api/admin/certificates/${certificateId}/blockchain/revoke`, {
    method: 'POST',
    headers: authHeaders({ 'Content-Type': 'application/json' }),
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json()
}
